Clarifying how handlers are handled (#77372)
ansible/ansible: docs/docsite/rst/user_guide/playbooks_handlers.rst

            name: httpd
            state: restarted

> 

Reviewer: ```suggestion
In this example playbook, the Apache server is restarted by the handler after all tasks complete in the play.
```

Reviewer: Feels like we should still have an explanation of what the first example is doing.

Notifying handlers
------------------

> Tasks can instruct one or more handlers to be executed using the ``notify`` keyword. The ``notify`` keyword can be applied to a task and accepts a list of handler names that would be notified on task change. Alternatively a string containing a single handler name can be supplied as well. The following example demonstrates how multiple handlers can be notified by a single task:

Reviewer: ```suggestion
Tasks can instruct one or more handlers to  execute using the ``notify`` keyword. The ``notify`` keyword can be applied to a task and accepts a list of handler names that  are notified on a task change. Alternately, a string containing a single handler name can be supplied as well. The following example demonstrates how multiple handlers can be notified by a single task:
```

          state: restarted
        listen: "restart web services"
      - name: Restart apache
        service:
          name: apache
          state: restarted
        listen: "restart web services"

Notifying ``restart web services`` topic will result in executing all handlers listening to that topic regardless of how those handlers are named.

This use makes it much easier to trigger multiple handlers. It also decouples handlers from their names, making it easier to share handlers among playbooks and roles (especially when using 3rd party roles from a shared source like Galaxy).

Each handler should have a globally unique name. If multiple handlers with the same name are defined only the last one defined can be notified via ``notify``, effectively shadowing all of the previous handlers with the same name. Alternatively handlers sharing the same name can all be notified and executed if they listen on the same topic by notifying that topic.

There is only one global scope for handlers (handler names and listen topics) regardless of where the handlers are defined. This includes handlers defined in roles as well.

Controlling when handlers run
-----------------------------

> By default, handlers run after all the tasks in a particular play have been completed. Notified handlers are executed automatically after each of the following sections: ``pre_tasks``, ``tasks`` and ``post_tasks``. This approach is efficient, because the handler only runs once, regardless of how many tasks notify it. For example, if multiple tasks update a configuration file and notify a handler to restart Apache, Ansible only bounces Apache once to avoid unnecessary restarts.

Reviewer 2: Should we say ``tasks``/``roles`` for this 2nd?  Not sure it would be clear to users of roles where their handlers fire.

Author: It is mentioned later in `Handlers in roles` but yes it would be better to mention that here as well.

Reviewer 3: also phrase that the 'multiple tasks' are 'in a section', handler A can be executed 4 times, if it gets notified in each section

Author: > also phrase that the 'multiple tasks' are 'in a section', handler A can be executed 4 times, if it gets notified in each section

It would be just 3 times because `roles` and `tasks` share the same implicit `flush_handers` task. Of course it might be more when `flush_handlers` is used within each of those sections, even multiple times.

I have that mentioned in the last sentence of `Controlling when handlers run` though, should that sentence be clarified more?

          name: apache
          state: restarted

> In the above example the handlers are executed on task change in the following order: ``Restart memcached``, ``Restart apache``. Handlers are executed in the order they are defined in the ``handlers`` section, not in the order listed in the ``notify`` statement. Notifying same handler multiple times will result in executing the handler only once regardless of how many tasks notify it. For example, if multiple tasks update a configuration file and notify a handler to restart Apache, Ansible only bounces Apache once to avoid unnecessary restarts.

Reviewer: ```suggestion
In the above example the handlers are executed on task change in the following order: ``Restart memcached``, ``Restart apache``. Handlers are executed in the order they are defined in the ``handlers`` section, not in the order listed in the ``notify`` statement. Notifying the same handler multiple times will result in executing the handler only once regardless of how many tasks notify it. For example, if multiple tasks update a configuration file and notify a handler to restart Apache, Ansible only bounces Apache once to avoid unnecessary restarts.
```

Naming handlers
---------------

Handlers must be named in order for tasks to be able to notify them using the ``notify`` keyword.

> Alternatively handlers can utilize the ``listen`` keyword. Using this handler keyword handlers can listen on topics that can group multiple handlers as follows:

Reviewer: ```suggestion
Alternately, handlers can utilize the ``listen`` keyword. Using this handler keyword, handlers can listen on topics that can group multiple handlers as follows:
```

.. code-block:: yaml

    tasks:
      - name: Restart everything
        command: echo "this task will restart the web services"
        notify: "restart web services"

    handlers:
      - name: Restart memcached
        service:
          name: memcached
          state: restarted
        listen: "restart web services"

      - name: Restart apache
        service:
          name: apache
          state: restarted
        listen: "restart web services"

> Notifying ``restart web services`` topic will result in executing all handlers listening to that topic regardless of how those handlers are named.

Reviewer: ```suggestion
Notifying the ``restart web services`` topic  results in executing all handlers listening to that topic regardless of how those handlers are named.
```

> This use makes it much easier to trigger multiple handlers. It also decouples handlers from their names, making it easier to share handlers among playbooks and roles (especially when using third-party roles from a shared source like Ansible Galaxy).

Reviewer: ```suggestion
This use makes it much easier to trigger multiple handlers. It also decouples handlers from their names, making it easier to share handlers among playbooks and roles (especially when using third-party roles from a shared source such as Ansible Galaxy).
```

> Each handler should have a globally unique name. If multiple handlers with the same name are defined only the last one defined can be notified via ``notify``, effectively shadowing all of the previous handlers with the same name. Alternatively handlers sharing the same name can all be notified and executed if they listen on the same topic by notifying that topic.

Reviewer: ```suggestion
Each handler should have a globally unique name. If multiple handlers are defined with the same name, only the last one defined is notified with ``notify``, effectively shadowing all of the previous handlers with the same name. Alternately handlers sharing the same name can all be notified and executed if they listen on the same topic by notifying that topic.
```

> There is only one global scope for handlers (handler names and listen topics) regardless of where the handlers are defined. This includes handlers defined in roles as well.

Reviewer: ```suggestion
There is only one global scope for handlers (handler names and listen topics) regardless of where the handlers are defined. This also includes handlers defined in roles.
```

The ``meta: flush_handlers`` task triggers any handlers that have been notified at that point in the play.

> Once handlers are executed, either automatically after each mentioned section or manually by ``flush_handlers`` meta task, they can be notified and run again in later sections of the play.

Reviewer: ```suggestion
Once handlers are executed, either automatically after each mentioned section or manually by  the ``flush_handlers`` meta task, they can be notified and run again in later sections of the play.
```

          name: "{{ web_service_name | default('httpd') }}"
          state: restarted

While handler names can contain a template, ``listen`` topics cannot.

Handlers in roles
-----------------

Handlers from roles are not just contained in their roles but rather inserted into global scope with all other handlers from a play. As such they can be used outside of the role they are defined in. It also means that their name can conflict with handlers from outside the role. To ensure that a handler from a role is notified as opposed to one from outside the role with the same name, notify the handler by using its name in the following form: ``role_name : handler_name``.

> Handlers notified within ``roles`` section are automatically flushed at the end of ``tasks`` section, but before any ``tasks`` handlers.

Reviewer: ```suggestion
Handlers notified within  the ``roles`` section are automatically flushed at the end of  the ``tasks`` section, but before any ``tasks`` handlers.
```

Includes and imports in handlers
--------------------------------

> Notifying a dynamic include like ``include_task`` as a handler will result in executing all tasks from within the include. It is not possible to notify a handler defined inside a dynamic include.

Reviewer: ```suggestion
Notifying a dynamic include such as ``include_task`` as a handler results in executing all tasks from within the include. It is not possible to notify a handler defined inside a dynamic include.
```

> Having a static include like ``import_task`` as a handler will result in that handler being effectively rewritten by handlers from within that import before the play execution. A static include itself cannot be notified, the tasks from withing that include, on the other hand, can be notified individually.

Reviewer: ```suggestion
Having a static include such as ``import_task`` as a handler results in that handler being effectively rewritten by handlers from within that import before the play execution. A static include itself cannot be notified, the tasks from withing that include, on the other hand, can be notified individually.
```
